File: QT/handler/ListEventHandler.cpp
#include "ListEventHandler.h"

#include <QDir>

#include <cmnDef.h>

#include <CLogger.h>
#include <assert.h>

#include "viewDefs.h"

namespace nsListEventHandler {
    const char* TAG = "ListEventHandler";
}

using namespace nsListEventHandler;

ListEventHandler::ListEventHandler( QObject *parent)
    : QObject(parent)
{

}

ListEventHandler::~ListEventHandler()
{
    disconnect( this, nullptr, nullptr, nullptr );
}

void ListEventHandler::connectObjectEvent( QObject * pObj )
{
    if( pObj == nullptr )
    {
        LOGE( TAG, "Object is null" );
        return;
    }

    QVariant vConnectEvent = pObj->property( PROP_B_CONNECT_EVENT );
    if( vConnectEvent.isNull() || vConnectEvent.isValid() == false
        || vConnectEvent.toBool() == false )
        return;

    QVariant vEventConnected = pObj->property( PROP_IS_EVENT_CONNECTED );
    if( vEventConnected.isNull() == false && vEventConnected.isValid()
            && vEventConnected.toBool() )
        return;

    QVariant vObjectType = pObj->property( "objectType" );
    if( vObjectType.isNull() || vObjectType.isValid() == false )
    {
        LOGE( TAG, "Object type is null" );
        return;
    }

    int nObjectType = vObjectType.toInt();
    if( nObjectType != static_cast<int>( ViewEnum::OBJECT_LIST_VIEW ) )
    {
        LOGE( TAG, "This object is not a button control. objectType: %d", nObjectType );
        return;
    }

    QVariant varEventConnectFlag = pObj->property( PROP_B_CONNECT_CLICKED_EVENT );
    if( varEventConnectFlag.isNull() == false && vEventConnected.isValid() && vEventConnected.toBool() )
        connect( pObj, SIGNAL( controlClickedEvent(int, int) ), this, SLOT( HandleControlClickedEvent(int, int) ), Qt::QueuedConnection );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_DOUBLE_CLICKED_EVENT );
    if( varEventConnectFlag.isNull() == false && vEventConnected.isValid() && vEventConnected.toBool() )
        connect( pObj, SIGNAL( controlDoubleClickedEvent(int, int) ), this, SLOT( HandleControlDoubleClickedEvent(int, int) ), Qt::QueuedConnection );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_ENTERED_EVENT );
    if( varEventConnectFlag.isNull() == false && varEventConnectFlag.isValid() && varEventConnectFlag.toBool() )
        connect( pObj, SIGNAL( controlEnteredEvent(int, int) ), this, SLOT( HandleControlEnteredEvent(int, int) ), Qt::QueuedConnection );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_EXITED_EVENT );
    if( varEventConnectFlag.isNull() == false && varEventConnectFlag.isValid() && varEventConnectFlag.toBool() )
        connect( pObj, SIGNAL( controlExitedEvent(int, int) ), this, SLOT( HandleControlExitedEvent(int, int) ), Qt::QueuedConnection );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_PRESS_AND_HOLD_EVENT );
    if( varEventConnectFlag.isNull() == false && vEventConnected.isValid() && vEventConnected.toBool() )
        connect( pObj, SIGNAL( controlPressAndHoldEvent(int, int) ), this, SLOT( HandleControlPressAndHoldEvent(int, int) ), Qt::QueuedConnection );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_PRESSED_EVENT );
    if( varEventConnectFlag.isNull() == false && vEventConnected.isValid() && vEventConnected.toBool() )
        connect( pObj, SIGNAL( controlPressedEvent(int, int) ), this, SLOT( HandleControlPressedEvent(int, int) ), Qt::QueuedConnection );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_RELEASED_EVENT );
    if( varEventConnectFlag.isNull() == false && vEventConnected.isValid() && vEventConnected.toBool() )
        connect( pObj, SIGNAL( controlReleasedEvent(int, int) ), this, SLOT( HandleControlReleasedEvent(int, int) ), Qt::QueuedConnection );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_CANCELED_EVENT );
    if( varEventConnectFlag.isNull() == false && vEventConnected.isValid() && vEventConnected.toBool() )
        connect( pObj, SIGNAL( controlCanceledEvent(int, int) ), this, SLOT( HandleControlCanceledEvent(int, int) ), Qt::QueuedConnection );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_WHEEL_EVENT );
    if( varEventConnectFlag.isNull() == false && vEventConnected.isValid() && vEventConnected.toBool() )
        connect( pObj, SIGNAL( controlWheelEvent(int, int, int, int) ), this, SLOT( HandleControlWheelEvent(int, int, int, int) ), Qt::QueuedConnection );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_HOVERED_EVENT );
    if( varEventConnectFlag.isNull() == false && vEventConnected.isValid() && vEventConnected.toBool() )
        connect( pObj, SIGNAL( controlHoveredEvent(int, int) ), this, SLOT( HandleControlHoveredEvent(int, int) ), Qt::QueuedConnection );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_FOCUSED_EVENT );
    if( varEventConnectFlag.isNull() == false && vEventConnected.isValid() && vEventConnected.toBool() )
        connect( pObj, SIGNAL( controlFocusedEvent(int, int) ), this, SLOT( HandleControlFocusedEvent(int, int) ), Qt::QueuedConnection );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_FOCUS_OUT_EVENT );
    if( varEventConnectFlag.isNull() == false && vEventConnected.isValid() && vEventConnected.toBool() )
        connect( pObj, SIGNAL( controlFocusOutEvent(int, int) ), this, SLOT( HandleControlFocusOutEvent(int, int) ), Qt::QueuedConnection );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_OPEN_EVENT );
    if( varEventConnectFlag.isNull() == false && vEventConnected.isValid() && vEventConnected.toBool() )
        connect( pObj, SIGNAL( controlOpenEvent(int, int) ), this, SLOT( HandleControlOpenEvent(int, int) ), Qt::QueuedConnection );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_CLOSED_EVENT );
    if( varEventConnectFlag.isNull() == false && vEventConnected.isValid() && vEventConnected.toBool() )
        connect( pObj, SIGNAL( controlClosedEvent(int, int) ), this, SLOT( HandleControlClosedEvent(int, int) ), Qt::QueuedConnection );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_DRAG_EVENT );
    if( varEventConnectFlag.isNull() == false && vEventConnected.isValid() && vEventConnected.toBool() )
        connect( pObj, SIGNAL( controlDragEvent( int, int, int, int ) ), this, SLOT( HandleControlDragEvent( int, int, int, int ) ), Qt::QueuedConnection );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_ENABLED_EVENT );
    if( varEventConnectFlag.isNull() == false && vEventConnected.isValid() && vEventConnected.toBool() )
        connect( pObj, SIGNAL( controlEnabledEvent(int, int) ), this, SLOT( HandleControlEnabledEvent(int, int) ), Qt::QueuedConnection );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_DISABLED_EVENT );
    if( varEventConnectFlag.isNull() == false && vEventConnected.isValid() && vEventConnected.toBool() )
        connect( pObj, SIGNAL( controlDisabledEvent(int, int) ), this, SLOT( HandleControlDisabledEvent(int, int) ), Qt::QueuedConnection );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_SELECTED_EVENT );
    if( varEventConnectFlag.isNull() == false && vEventConnected.isValid() && vEventConnected.toBool() )
        connect( pObj, SIGNAL( controlSelectedEvent(int, int) ), this, SLOT( HandleControlSelectedEvent(int, int) ), Qt::QueuedConnection );

    pObj->setProperty( PROP_IS_EVENT_CONNECTED, true );
}

void ListEventHandler::disconnectObjectEvent( QObject * pObj )
{
    if( pObj == nullptr )
    {
        LOGE( TAG, "Object is null" );
        return;
    }

    QVariant vConnectEvent = pObj->property( PROP_B_CONNECT_EVENT );
    if( vConnectEvent.isNull() || vConnectEvent.isValid() == false
        || vConnectEvent.toBool() == false )
        return;

    QVariant vEventConnected = pObj->property( PROP_IS_EVENT_CONNECTED );
    if( vEventConnected.isNull() || vEventConnected.isValid() == false
        || vEventConnected.toBool() == false )
        return;

    QVariant vObjectType = pObj->property( "objectType" );
    if( vObjectType.isNull() || vObjectType.isValid() == false )
    {
        LOGE( TAG, "Object type is null" );
        return;
    }

    int nObjectType = vObjectType.toInt();
    if( nObjectType != static_cast<int>( ViewEnum::OBJECT_LIST_VIEW ) )
    {
        LOGE( TAG, "This object is not a button control. objectType: %d", nObjectType );
        return;
    }

    QVariant varEventConnectFlag = pObj->property( PROP_B_CONNECT_CLICKED_EVENT );
    if( varEventConnectFlag.isNull() == false && varEventConnectFlag.isValid() && varEventConnectFlag.toBool() )
        disconnect( pObj, SIGNAL( controlClickedEvent(int, int) ), this, SLOT( HandleControlClickedEvent(int, int) ) );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_DOUBLE_CLICKED_EVENT );
    if( varEventConnectFlag.isNull() == false && varEventConnectFlag.isValid() && varEventConnectFlag.toBool() )
        disconnect( pObj, SIGNAL( controlDoubleClickedEvent(int, int) ), this, SLOT( HandleControlDoubleClickedEvent(int, int) ) );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_ENTERED_EVENT );
    if( varEventConnectFlag.isNull() == false && varEventConnectFlag.isValid() && varEventConnectFlag.toBool() )
        disconnect( pObj, SIGNAL( controlEnteredEvent(int, int) ), this, SLOT( HandleControlEnteredEvent(int, int) ) );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_EXITED_EVENT );
    if( varEventConnectFlag.isNull() == false && varEventConnectFlag.isValid() && varEventConnectFlag.toBool() )
        disconnect( pObj, SIGNAL( controlExitedEvent(int, int) ), this, SLOT( HandleControlExitedEvent(int, int) ) );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_PRESS_AND_HOLD_EVENT );
    if( varEventConnectFlag.isNull() == false && varEventConnectFlag.isValid() && varEventConnectFlag.toBool() )
        disconnect( pObj, SIGNAL( controlPressAndHoldEvent(int, int) ), this, SLOT( HandleControlPressAndHoldEvent(int, int) ) );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_PRESSED_EVENT );
    if( varEventConnectFlag.isNull() == false && varEventConnectFlag.isValid() && varEventConnectFlag.toBool() )
        disconnect( pObj, SIGNAL( controlPressedEvent(int, int) ), this, SLOT( HandleControlPressedEvent(int, int) ) );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_RELEASED_EVENT );
    if( varEventConnectFlag.isNull() == false && varEventConnectFlag.isValid() && varEventConnectFlag.toBool() )
        disconnect( pObj, SIGNAL( controlReleasedEvent(int, int) ), this, SLOT( HandleControlReleasedEvent(int, int) ) );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_CANCELED_EVENT );
    if( varEventConnectFlag.isNull() == false && varEventConnectFlag.isValid() && varEventConnectFlag.toBool() )
        disconnect( pObj, SIGNAL( controlCanceledEvent(int, int) ), this, SLOT( HandleControlCanceledEvent(int, int) ) );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_WHEEL_EVENT );
    if( varEventConnectFlag.isNull() == false && varEventConnectFlag.isValid() && varEventConnectFlag.toBool() )
        disconnect( pObj, SIGNAL( controlWheelEvent(int, int, int, int) ), this, SLOT( HandleControlWheelEvent(int, int, int, int) ) );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_HOVERED_EVENT );
    if( varEventConnectFlag.isNull() == false && varEventConnectFlag.isValid() && varEventConnectFlag.toBool() )
        disconnect( pObj, SIGNAL( controlHoveredEvent(int, int) ), this, SLOT( HandleControlHoveredEvent(int, int) ) );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_FOCUSED_EVENT );
    if( varEventConnectFlag.isNull() == false && varEventConnectFlag.isValid() && varEventConnectFlag.toBool() )
        disconnect( pObj, SIGNAL( controlFocusedEvent(int, int) ), this, SLOT( HandleControlFocusedEvent(int, int) ) );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_FOCUS_OUT_EVENT );
    if( varEventConnectFlag.isNull() == false && varEventConnectFlag.isValid() && varEventConnectFlag.toBool() )
        disconnect( pObj, SIGNAL( controlFocusOutEvent(int, int) ), this, SLOT( HandleControlFocusOutEvent(int, int) ) );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_OPEN_EVENT );
    if( varEventConnectFlag.isNull() == false && varEventConnectFlag.isValid() && varEventConnectFlag.toBool() )
        disconnect( pObj, SIGNAL( controlOpenEvent(int, int) ), this, SLOT( HandleControlOpenEvent(int, int) ) );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_CLOSED_EVENT );
    if( varEventConnectFlag.isNull() == false && varEventConnectFlag.isValid() && varEventConnectFlag.toBool() )
        disconnect( pObj, SIGNAL( controlClosedEvent(int, int) ), this, SLOT( HandleControlClosedEvent(int, int) ) );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_DRAG_EVENT );
    if( varEventConnectFlag.isNull() == false && varEventConnectFlag.isValid() && varEventConnectFlag.toBool() )
        disconnect( pObj, SIGNAL( controlDragEvent( int, int, int, int ) ), this, SLOT( HandleControlDragEvent( int, int, int, int ) ) );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_ENABLED_EVENT );
    if( varEventConnectFlag.isNull() == false && varEventConnectFlag.isValid() && varEventConnectFlag.toBool() )
        disconnect( pObj, SIGNAL( controlEnabledEvent(int, int) ), this, SLOT( HandleControlEnabledEvent(int, int) ) );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_DISABLED_EVENT );
    if( varEventConnectFlag.isNull() == false && varEventConnectFlag.isValid() && varEventConnectFlag.toBool() )
        disconnect( pObj, SIGNAL( controlDisabledEvent(int, int) ), this, SLOT( HandleControlDisabledEvent(int, int) ) );

    varEventConnectFlag = pObj->property( PROP_B_CONNECT_SELECTED_EVENT );
    if( varEventConnectFlag.isNull() == false && varEventConnectFlag.isValid() && varEventConnectFlag.toBool() )
        disconnect( pObj, SIGNAL( controlSelectedEvent(int, int) ), this, SLOT( HandleControlSelectedEvent(int, int) ) );

    pObj->setProperty( PROP_IS_EVENT_CONNECTED, false );
}

void ListEventHandler::HandleControlClickedEvent( int row, int col )
{
    QObject* pObj = sender();

    const QString& strObjName = pObj->objectName();

    LOGD( TAG, "strObjName: %s", strObjName.toStdString().c_str() );
}

void ListEventHandler::HandleControlDoubleClickedEvent( int row, int col )
{
    QObject* pObj = sender();

    const QString& strObjName = pObj->objectName();

    LOGD( TAG, "strObjName: %s", strObjName.toStdString().c_str() );
}

void ListEventHandler::HandleControlEnteredEvent( int row, int col )
{
    QObject* pObj = sender();

    const QString& strObjName = pObj->objectName();

    LOGD( TAG, "strObjName: %s", strObjName.toStdString().c_str() );
}

void ListEventHandler::HandleControlExitedEvent( int row, int col )
{
    QObject* pObj = sender();

    const QString& strObjName = pObj->objectName();

    LOGD( TAG, "strObjName: %s", strObjName.toStdString().c_str() );
}

void ListEventHandler::HandleControlPressAndHoldEvent( int row, int col )
{
    QObject* pObj = sender();

    const QString& strObjName = pObj->objectName();

    LOGD( TAG, "strObjName: %s", strObjName.toStdString().c_str() );
}

void ListEventHandler::HandleControlPressedEvent( int row, int col )
{
    QObject* pObj = sender();

    const QString& strObjName = pObj->objectName();

    LOGD( TAG, "strObjName: %s", strObjName.toStdString().c_str() );
}

void ListEventHandler::HandleControlReleasedEvent( int row, int col )
{
    QObject* pObj = sender();

    const QString& strObjName = pObj->objectName();

    LOGD( TAG, "strObjName: %s", strObjName.toStdString().c_str() );
}

void ListEventHandler::HandleControlCanceledEvent( int row, int col )
{
    QObject* pObj = sender();

    const QString& strObjName = pObj->objectName();

    LOGD( TAG, "strObjName: %s", strObjName.toStdString().c_str() );
}

void ListEventHandler::HandleControlWheelEvent( int row, int col, int x, int y )
{
    QObject* pObj = sender();

    const QString& strObjName = pObj->objectName();

    LOGD( TAG, "strObjName: %s", strObjName.toStdString().c_str() );
}

void ListEventHandler::HandleControlHoveredEvent( int row, int col )
{
    QObject* pObj = sender();

    const QString& strObjName = pObj->objectName();

    LOGD( TAG, "strObjName: %s", strObjName.toStdString().c_str() );
}

void ListEventHandler::HandleControlFocusedEvent( int row, int col )
{
    QObject* pObj = sender();

    const QString& strObjName = pObj->objectName();

    LOGD( TAG, "strObjName: %s", strObjName.toStdString().c_str() );
}

void ListEventHandler::HandleControlFocusOutEvent( int row, int col )
{
    QObject* pObj = sender();

    const QString& strObjName = pObj->objectName();

    LOGD( TAG, "strObjName: %s", strObjName.toStdString().c_str() );
}

void ListEventHandler::HandleControlOpenEvent( int row, int col )
{
    QObject* pObj = sender();

    const QString& strObjName = pObj->objectName();

    LOGD( TAG, "strObjName: %s", strObjName.toStdString().c_str() );
}

void ListEventHandler::HandleControlClosedEvent( int row, int col )
{
    QObject* pObj = sender();

    const QString& strObjName = pObj->objectName();

    LOGD( TAG, "strObjName: %s", strObjName.toStdString().c_str() );
}

void ListEventHandler::HandleControlDragEvent( int row, int col, int x, int y )
{
    QObject* pObj = sender();

    const QString& strObjName = pObj->objectName();

    LOGD( TAG, "strObjName: %s", strObjName.toStdString().c_str() );
}

void ListEventHandler::HandleControlEnabledEvent( int row, int col )
{
    QObject* pObj = sender();

    const QString& strObjName = pObj->objectName();

    LOGD( TAG, "strObjName: %s", strObjName.toStdString().c_str() );
}

void ListEventHandler::HandleControlDisabledEvent( int row, int col )
{
    QObject* pObj = sender();

    const QString& strObjName = pObj->objectName();

    LOGD( TAG, "strObjName: %s", strObjName.toStdString().c_str() );
}

void ListEventHandler::HandleControlSelectedEvent( int row, int col )
{
    QObject* pObj = sender();

    const QString& strObjName = pObj->objectName();

    LOGD( TAG, "strObjName: %s", strObjName.toStdString().c_str() );
}

File: python/module/ViewControlModule/handler/ListEventHandler.py
# This is a sample Python script.

# Press Shift+F10 to execute it or replace it with your code.
# Press Double Shift to search everywhere for classes, files, tool windows, actions, and settings.

from PyQt5 import QtCore
from PyQt5.QtCore import QObject
from PyQt5.QtCore import pyqtSignal, pyqtSlot, Qt

from python.src.define.viewDefs import *
from python.src.utils.logger import getLogger

class ListEventHandler(QtCore.QObject):

    def __init__(self, parent=None):
        super().__init__(parent)

    def connectObjectEvent(self, obj:QtCore.QObject):
        if not obj:
            getLogger().error('object is null')

        bConnectEvent = obj.property(PROP_B_CONNECT_EVENT)
        if not bConnectEvent:
            return

        nEventConnected = obj.property(PROP_IS_EVENT_CONNECTED)
        if nEventConnected != None and nEventConnected == True:
            return

        nObjectType = obj.property('objectType')
        if not nObjectType:
            getLogger().error('Object type is null')
            return

        if nObjectType != int(ViewEnum.eObjectType.OBJECT_LIST_VIEW):
            getLogger().error('This object is not a list control. objectType: %d' % nObjectType)
            return

        bEventConnectFlag = obj.property(PROP_B_CONNECT_CLICKED_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlClickedEvent.connect(self.HandleControlClickedEvent, Qt.DirectConnection)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_DOUBLE_CLICKED_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlDoubleClickedEvent.connect(self.HandleControlDoubleClickedEvent, Qt.DirectConnection)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_ENTERED_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlEnteredEvent.connect(self.HandleControlEnteredEvent, Qt.DirectConnection)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_EXITED_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlExitedEvent.connect(self.HandleControlExitedEvent, Qt.DirectConnection)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_PRESS_AND_HOLD_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlPressAndHoldEvent.connect(self.HandleControlPressAndHoldEvent, Qt.DirectConnection)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_PRESSED_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlPressedEvent.connect(self.HandleControlPressedEvent, Qt.DirectConnection)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_RELEASED_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlReleasedEvent.connect(self.HandleControlReleasedEvent, Qt.DirectConnection)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_CANCELED_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlCanceledEvent.connect(self.HandleControlCanceledEvent, Qt.DirectConnection)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_WHEEL_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlWheelEvent.connect(self.HandleControlWheelEvent, Qt.DirectConnection)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_HOVERED_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlHoveredEvent.connect(self.HandleControlHoveredEvent, Qt.DirectConnection)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_FOCUSED_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlFocusedEvent.connect(self.HandleControlFocusedEvent, Qt.DirectConnection)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_FOCUS_OUT_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlFocusOutEvent.connect(self.HandleControlFocusOutEvent, Qt.DirectConnection)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_OPEN_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlOpenEvent.connect(self.HandleControlOpenEvent, Qt.DirectConnection)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_CLOSED_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlClosedEvent.connect(self.HandleControlClosedEvent, Qt.DirectConnection)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_DRAG_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlDragEvent.connect(self.HandleControlDragEvent, Qt.DirectConnection)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_ENABLED_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlEnabledEvent.connect(self.HandleControlEnabledEvent, Qt.DirectConnection)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_DISABLED_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlDisabledEvent.connect(self.HandleControlDisabledEvent, Qt.DirectConnection)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_SELECTED_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlSelectedEvent.connect(self.HandleControlSelectedEvent, Qt.DirectConnection)

        obj.setProperty(PROP_IS_EVENT_CONNECTED, True)

    def disconnectObjectEvent(self, obj:QtCore.QObject):
        if not obj:
            getLogger().error('object is null')

        bConnectEvent = obj.property(PROP_B_CONNECT_EVENT)
        if not bConnectEvent:
            return

        nEventConnected = obj.property(PROP_IS_EVENT_CONNECTED)
        if not nEventConnected:
            return

        nObjectType = obj.property('objectType')
        if not nObjectType:
            getLogger().error('Object type is null')
            return

        if nObjectType != int( ViewEnum.eObjectType.OBJECT_LIST_VIEW ):
            getLogger().error('This object is not a list control. objectType: %d' % nObjectType)
            return

        bEventConnectFlag = obj.property( PROP_B_CONNECT_CLICKED_EVENT )
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlClickedEvent.disconnect(self.HandleControlClickedEvent)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_DOUBLE_CLICKED_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlDoubleClickedEvent.disconnect(self.HandleControlDoubleClickedEvent)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_ENTERED_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlEnteredEvent.disconnect(self.HandleControlEnteredEvent)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_EXITED_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlExitedEvent.disconnect(self.HandleControlExitedEvent)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_PRESS_AND_HOLD_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlPressAndHoldEvent.disconnect(self.HandleControlPressAndHoldEvent)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_PRESSED_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlPressedEvent.disconnect(self.HandleControlPressedEvent)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_RELEASED_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlReleasedEvent.disconnect(self.HandleControlReleasedEvent)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_CANCELED_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlCanceledEvent.disconnect(self.HandleControlCanceledEvent)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_WHEEL_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlWheelEvent.disconnect(self.HandleControlWheelEvent)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_HOVERED_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlHoveredEvent.disconnect(self.HandleControlHoveredEvent)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_FOCUSED_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlFocusedEvent.disconnect(self.HandleControlFocusedEvent)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_FOCUS_OUT_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlFocusOutEvent.disconnect(self.HandleControlFocusOutEvent)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_OPEN_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlOpenEvent.disconnect(self.HandleControlOpenEvent)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_CLOSED_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlClosedEvent.disconnect(self.HandleControlClosedEvent)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_DRAG_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlDragEvent.disconnect(self.HandleControlDragEvent)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_ENABLED_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlEnabledEvent.disconnect(self.HandleControlEnabledEvent)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_DISABLED_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlDisabledEvent.disconnect(self.HandleControlDisabledEvent)

        bEventConnectFlag = obj.property(PROP_B_CONNECT_SELECTED_EVENT)
        if bEventConnectFlag != None and bEventConnectFlag == True:
            obj.controlSelectedEvent.disconnect(self.HandleControlSelectedEvent)

        obj.setProperty(PROP_IS_EVENT_CONNECTED, False)

    @QtCore.pyqtSlot(int, int)
    def HandleControlClickedEvent(self, row, col):
        obj = self.sender()
        if not obj:
            getLogger().error('obj is Null')
            return

        strObjName = obj.objectName()
        getLogger().debug(f'strObjName: {strObjName}')

    @QtCore.pyqtSlot(int, int)
    def HandleControlDoubleClickedEvent(self, row, col):
        obj = self.sender()
        if not obj:
            getLogger().error('obj is Null')
            return

        strObjName = obj.objectName()
        getLogger().debug(f'strObjName: {strObjName}')

    @QtCore.pyqtSlot(int, int)
    def HandleControlEnteredEvent(self, row, col):
        obj = self.sender()
        if not obj:
            getLogger().error('obj is Null')
            return

        strObjName = obj.objectName()
        getLogger().debug(f'strObjName: {strObjName}')

    @QtCore.pyqtSlot(int, int)
    def HandleControlExitedEvent(self, row, col):
        obj = self.sender()
        if not obj:
            getLogger().error('obj is Null')
            return

        strObjName = obj.objectName()
        getLogger().debug(f'strObjName: {strObjName}')

    @QtCore.pyqtSlot(int, int)
    def HandleControlPressAndHoldEvent(self, row, col):
        obj = self.sender()
        if not obj:
            getLogger().error('obj is Null')
            return

        strObjName = obj.objectName()
        getLogger().debug(f'strObjName: {strObjName}')

    @QtCore.pyqtSlot(int, int)
    def HandleControlPressedEvent(self, row, col):
        obj = self.sender()
        if not obj:
            getLogger().error('obj is Null')
            return

        strObjName = obj.objectName()
        getLogger().debug(f'strObjName: {strObjName}')

    @QtCore.pyqtSlot(int, int)
    def HandleControlReleasedEvent(self, row, col):
        obj = self.sender()
        if not obj:
            getLogger().error('obj is Null')
            return

        strObjName = obj.objectName()
        getLogger().debug(f'strObjName: {strObjName}')

    @QtCore.pyqtSlot(int, int)
    def HandleControlCanceledEvent(self, row, col):
        obj = self.sender()
        if not obj:
            getLogger().error('obj is Null')
            return

        strObjName = obj.objectName()
        getLogger().debug(f'strObjName: {strObjName}')

    @QtCore.pyqtSlot(int, int, int, int)
    def HandleControlWheelEvent(self, row, col, x, y):
        obj = self.sender()
        if not obj:
            getLogger().error('obj is Null')
            return

        strObjName = obj.objectName()
        getLogger().debug(f'strObjName: {strObjName}')

    @QtCore.pyqtSlot(int, int)
    def HandleControlHoveredEvent(self, row, col):
        obj = self.sender()
        if not obj:
            getLogger().error('obj is Null')
            return

        strObjName = obj.objectName()
        getLogger().debug(f'strObjName: {strObjName}')

    @QtCore.pyqtSlot(int, int)
    def HandleControlFocusedEvent(self, row, col):
        obj = self.sender()
        if not obj:
            getLogger().error('obj is Null')
            return

        strObjName = obj.objectName()
        getLogger().debug(f'strObjName: {strObjName}')

    @QtCore.pyqtSlot(int, int)
    def HandleControlFocusOutEvent(self, row, col):
        obj = self.sender()
        if not obj:
            getLogger().error('obj is Null')
            return

        strObjName = obj.objectName()
        getLogger().debug(f'strObjName: {strObjName}')

    @QtCore.pyqtSlot(int, int)
    def HandleControlOpenEvent(self, row, col):
        obj = self.sender()
        if not obj:
            getLogger().error('obj is Null')
            return

        strObjName = obj.objectName()
        getLogger().debug(f'strObjName: {strObjName}')

    @QtCore.pyqtSlot(int, int)
    def HandleControlClosedEvent(self, row, col):
        obj = self.sender()
        if not obj:
            getLogger().error('obj is Null')
            return

        strObjName = obj.objectName()
        getLogger().debug(f'strObjName: {strObjName}')

    @QtCore.pyqtSlot(int, int, int ,int)
    def HandleControlDragEvent(self, row, col, x, y):
        obj = self.sender()
        if not obj:
            getLogger().error('obj is Null')
            return

        strObjName = obj.objectName()
        getLogger().debug(f'strObjName: {strObjName}')

    @QtCore.pyqtSlot(int, int)
    def HandleControlEnabledEvent(self, row, col):
        obj = self.sender()
        if not obj:
            getLogger().error('obj is Null')
            return

        strObjName = obj.objectName()
        getLogger().debug(f'strObjName: {strObjName}')

    @QtCore.pyqtSlot(int, int)
    def HandleControlDisabledEvent(self, row, col):
        obj = self.sender()
        if not obj:
            getLogger().error('obj is Null')
            return

        strObjName = obj.objectName()
        getLogger().debug(f'strObjName: {strObjName}')

    @QtCore.pyqtSlot(int, int)
    def HandleControlSelectedEvent(self, row, col):
        obj = self.sender()
        if not obj:
            getLogger().error('obj is Null')
            return

        strObjName = obj.objectName()
        getLogger().debug(f'strObjName: {strObjName}')

# See PyCharm help at https://www.jetbrains.com/help/pycharm/
